File: src/main/java/ml/wonwoo/springdatashowcase/support/MonetaryAmountAttributeConverter.java
package ml.wonwoo.springdatashowcase.support;

import java.util.Locale;
import javax.money.MonetaryAmount;
import javax.money.format.MonetaryAmountFormat;
import javax.money.format.MonetaryFormats;
import javax.persistence.AttributeConverter;
import javax.persistence.Converter;
import org.javamoney.moneta.Money;

@Converter(autoApply = true)
public class MonetaryAmountAttributeConverter implements AttributeConverter<MonetaryAmount, String> {

    private static final MonetaryAmountFormat FORMAT = MonetaryFormats.getAmountFormat(Locale.ROOT);

    @Override
    public String convertToDatabaseColumn(MonetaryAmount amount) {
        return amount == null ? null : amount.toString();
    }

    @Override
    public MonetaryAmount convertToEntityAttribute(String source) {

        if (source == null) {
            return null;
        }

        try {
            return Money.parse(source);
        } catch (RuntimeException e) {
            return Money.parse(source, FORMAT);
        }
    }
}

File: kotlin/src/main/kotlin/ml/wonwoo/springdatashowcase/support/MonetaryAmountAttributeConverter.kt
package ml.wonwoo.springdatashowcase.support

import java.util.Locale
import javax.money.MonetaryAmount
import javax.money.format.MonetaryFormats
import javax.persistence.AttributeConverter
import javax.persistence.Converter

@Converter(autoApply = true)
class MonetaryAmountAttributeConverter : AttributeConverter<MonetaryAmount, String> {

    private val format = MonetaryFormats.getAmountFormat(Locale.ROOT)

    override fun convertToDatabaseColumn(amount: MonetaryAmount?): String? {
        return amount?.toString()
    }

    override fun convertToEntityAttribute(source: String?): MonetaryAmount? {

        return if (source == null) null else format.parse(source)

    }

}